### PR TITLE
New Resource: aws_lambda_layer_version_permission

### LINKED ISSUE #8595: Add lambda_layer_version_permission resource
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Related to #6651, specifically [this comment](https://github.com/terraform-providers/terraform-provider-aws/issues/6651#issuecomment-454431427) a new resource (`aws_lambda_layer_version_permission`) needs to be created to support cross-account access to lambda layer versions.

### New or Affected Resource(s)

* `aws_lambda_layer_version_permission`

### Potential Terraform Configuration

```hcl
resource "aws_lambda_layer_version_permission" "example" {
  statement_id  = "allow org access"
  action        = "lambda:GetLayerVersion"
  layer_name    = "${aws_lambda_layer_version.example.layer_name}"
  principal     = "*"
  version       = "${aws_lambda_layer_version.example.version}"
  organization  = "o-123456789"
}
```

### References

* #6651

### PR DESCRIPTION
**Pull request by Avolynsk**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #8595

Release note for CHANGELOG:
```
* **New Resource:** `aws_lambda_layer_version_permission`
```

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSLambdaLayerVersionPermission'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSLambdaLayerVersionPermission -timeout 120m
=== RUN   TestAccAWSLambdaLayerVersionPermission_all
=== PAUSE TestAccAWSLambdaLayerVersionPermission_all
=== RUN   TestAccAWSLambdaLayerVersionPermission_org
=== PAUSE TestAccAWSLambdaLayerVersionPermission_org
=== RUN   TestAccAWSLambdaLayerVersionPermission_account
=== PAUSE TestAccAWSLambdaLayerVersionPermission_account
=== RUN   TestAccAWSLambdaLayerVersionPermission_disappears
=== PAUSE TestAccAWSLambdaLayerVersionPermission_disappears
=== CONT  TestAccAWSLambdaLayerVersionPermission_all
=== CONT  TestAccAWSLambdaLayerVersionPermission_disappears
=== CONT  TestAccAWSLambdaLayerVersionPermission_account
=== CONT  TestAccAWSLambdaLayerVersionPermission_org
    TestAccAWSLambdaLayerVersionPermission_disappears: resource_aws_lambda_layer_version_permission_test.go:160: [INFO] Got non-empty plan, as expected
--- PASS: TestAccAWSLambdaLayerVersionPermission_disappears (45.52s)
--- PASS: TestAccAWSLambdaLayerVersionPermission_all (53.67s)
--- PASS: TestAccAWSLambdaLayerVersionPermission_org (53.70s)
--- PASS: TestAccAWSLambdaLayerVersionPermission_account (53.83s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	56.582s
```

### CODE REVIEW

**Comment by Avolynsk** [comment]
WIP, will add tests soon.

**Comment by CliffHoogervorst** [comment]
When is this expected to be merged??

**Comment by Avolynsk** [comment]
@bflad could you pls take a look? Sorry for bothering you, just wondering if I've missed some contributing rules for my PR

**Comment by CliffHoogervorst** [comment]
@bflad  can you please check ??

File: aws/resource_aws_lambda_layer_version_permission.go
Comment (DrFaust92) [inline aws/resource_aws_lambda_layer_version_permission.go:31]: add 
`ValidateFunc: validateArn,`

File: aws/resource_aws_lambda_layer_version_permission.go
Comment (DrFaust92) [inline aws/resource_aws_lambda_layer_version_permission.go:86]: can the above vars be inside the `AddLayerVersionPermissionInput` block?

like so:
```
	params := &lambda.AddLayerVersionPermissionInput{
		LayerName:     aws.String(d.Get("layer_arn").(string)),
... other arguments
	}
```

File: aws/resource_aws_lambda_layer_version_permission.go
Comment (DrFaust92) [inline aws/resource_aws_lambda_layer_version_permission.go:98]: d.SetId(fmt.sprintf("%s:%s",aws.StringValue(layerArn),strconv.FormatInt(aws.Int64Value(int(layerVersion)), 10))

File: aws/resource_aws_lambda_layer_version_permission.go
Comment (DrFaust92) [inline aws/resource_aws_lambda_layer_version_permission.go:99]: i think this can be ommited on create

File: aws/resource_aws_lambda_layer_version_permission.go
Comment (DrFaust92) [inline aws/resource_aws_lambda_layer_version_permission.go:21]: change function to align with convention: *Create, *Read, *Delete

File: aws/resource_aws_lambda_layer_version_permission_test.go
Comment (DrFaust92) [inline aws/resource_aws_lambda_layer_version_permission_test.go:67]: `rName := acctest.RandomWithPrefix("tf-acc-test")` is the standard for test resource names

File: aws/resource_aws_lambda_layer_version_permission_test.go
Comment (DrFaust92) [inline aws/resource_aws_lambda_layer_version_permission_test.go:78]: use `resourceName`

File: aws/resource_aws_lambda_layer_version_permission_test.go
Comment (DrFaust92) [inline aws/resource_aws_lambda_layer_version_permission_test.go:107]: same as above

File: aws/resource_aws_lambda_layer_version_permission_test.go
Comment (DrFaust92) [inline aws/resource_aws_lambda_layer_version_permission_test.go:137]: same as above

File: aws/resource_aws_lambda_layer_version_permission_test.go
Comment (DrFaust92) [inline aws/resource_aws_lambda_layer_version_permission_test.go:81]: lets also add a test of comparing the `layer_arn` attribute with `arn` of layer:
`resource.TestCheckResourceAttrPair(resourceName, "layer_arn", "aws_lambda_layer_version.lambda_layer", "arn"),`

**Comment by Jorge-Rodriguez** [comment]
@Avolynsk @DrFaust92 do you have an estimate on how much more work/time would it take for this PR to be merged?

**Comment by sunilkumarmohanty** [comment]
This PR has not been updated for 6 months now. Should we create a duplicate PR with the changes?
pinging @bflad

**Comment by teamterraform** [comment]
## Notification of Recent and Upcoming Changes to Contributions

Thank you for this contribution! There have been a few recent development changes that affect this pull request. We apologize for the inconvenience, especially if there have been long review delays up until now. Please note that this is automated message from an unmonitored account. See the [FAQ](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md) for additional information on the maintainer team and review prioritization.

If you are unable to complete these updates, please leave a comment for the community and maintainers so someone can potentially continue the work. The maintainers will encourage other contributors to use the existing contribution as the base for additional changes as appropriate. Otherwise, contributions that do not receive updated code or comments from the original contributor may be closed in the future so the maintainers can focus on active items.

For the most up to date information about Terraform AWS Provider development, see the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md). Additional technical debt changes can be tracked with the [`technical-debt` label on issues](https://github.com/terraform-providers/terraform-provider-aws/issues?q=is%3Aopen+is%3Aissue+label%3Atechnical-debt).

As part of updating a pull request with these changes, the most current unit testing and linting will run. These may report issues that were not previously reported.

### Terraform 0.12 Syntax

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/8950
Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14417

Version 3 and later of the Terraform AWS Provider, which all existing contributions would potentially be added, only supports Terraform 0.12 and later. Certain syntax elements of Terraform 0.11 and earlier show deprecation warnings during runs with Terraform 0.12. Documentation and test configurations, such as those including deprecated string interpolations (`some_attribute = "${aws_service_thing.example.id}"`) should be updated to the newer syntax (`some_attribute = aws_service_thing.example.id`). Contribution testing will automatically fail on older syntax in the near future. Please see the referenced issues for additional information.

### Action Required: Terraform Plugin SDK Version 2

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14551

The Terraform AWS Provider has been upgraded to the latest version of the Terraform Plugin SDK. Generally, most changes to contributions should only involve updating Go import paths in source code files. Please see the referenced issue for additional information.

### Action Required: Removal of website/aws.erb File

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14712

Any changes to the `website/aws.erb` file are no longer necessary and should be removed from this contribution to prevent merge issues in the near future when the file is removed from the repository. Please see the referenced issue for additional information.

### Upcoming Change of Git Branch Naming

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14292

Development environments will need their upstream Git branch updated from `master` to `main` in the near future. Please see the referenced issue for additional information and scheduling.

### Upcoming Change of GitHub Organization

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14715

This repository will be migrating from https://github.com/terraform-providers/terraform-provider-aws to https://github.com/hashicorp/terraform-provider-aws. No practitioner or developer action is anticipated and most GitHub functionality will automatically redirect to the new location. Go import paths including `terraform-providers` can remain for now. Please see the referenced issue for additional information and scheduling.

**Comment by Jorge-Rodriguez** [comment]
@maryelizbeth @bflad @gdavison @angie44 @breathingdust @ksatirli
Any news on this PR? @sunilkumarmohanty seems willing to get it up to snuff if necessary.
Personally, I would like to see this merged soon.

**Comment by bflad** [comment]
@Jorge-Rodriguez if there is no response from the original contributor, @Avolynsk, the maintainers will recommend a new pull request be submitted that contains changes on top of this one. It cannot be merged as-is.

**Comment by Avolynsk** [comment]
Hi guys! Sorry for the delay, I'll try to update the PR next week.

**Comment by sunilkumarmohanty** [comment]
Thanks @Avolynsk . However, if you are busy, please let me know. I can help with finishing the PR.

File: aws/resource_aws_lambda_layer_version_permission.go
Comment (Avolynsk) [reply aws/resource_aws_lambda_layer_version_permission.go]: added

File: aws/resource_aws_lambda_layer_version_permission.go
Comment (Avolynsk) [reply aws/resource_aws_lambda_layer_version_permission.go]: vars moved inside

File: aws/resource_aws_lambda_layer_version_permission.go
Comment (Avolynsk) [reply aws/resource_aws_lambda_layer_version_permission.go]: updated

File: aws/resource_aws_lambda_layer_version_permission.go
Comment (Avolynsk) [reply aws/resource_aws_lambda_layer_version_permission.go]: functions names have been updated

File: aws/resource_aws_lambda_layer_version_permission_test.go
Comment (Avolynsk) [reply aws/resource_aws_lambda_layer_version_permission_test.go]: Updated

File: aws/resource_aws_lambda_layer_version_permission_test.go
Comment (Avolynsk) [reply aws/resource_aws_lambda_layer_version_permission_test.go:81]: Test added.

File: aws/resource_aws_lambda_layer_version_permission_test.go
Comment (Avolynsk) [reply aws/resource_aws_lambda_layer_version_permission_test.go]: updated

File: aws/resource_aws_lambda_layer_version_permission.go
Comment (Avolynsk) [reply aws/resource_aws_lambda_layer_version_permission.go]: removed

**Comment by Avolynsk** [comment]
hi @DrFaust92 ! I've fixed all things suggested by you during the review. The only thing that I'm not sure about is `disappears` test, could you pls take a look?

And during my work, I've merged an actual master branch into my own to get `make build/tests` works again, and seems this is causing `CHANGELOG Checks / Comment (pull_request)` check failure. Also, I've noticed that now my PR shows **too many** changed files, so now I don't know how to figure this out.
Could you pls help me with this?

**Comment by unix-system** [comment]
I can't see the issue with the PR that caused the test fail - perhaps rerun the test?

Keen for this to get merged, as it's a blocker for me. Let me know if you could use any help to push it over the line 🙂

**Comment by anGie44** [comment]
Hi @Avolynsk , I'm going to look and see if I can rebase this on my end to a previous commit before `main` was merged in, will probably result in some check/format errors which you fixed more recently but I can have a look 👍

**Comment by BrandonALXEllisSS** [comment]
Is this ever going to get merged?

**Comment by breathingdust** [comment]
Hi all 👋  Just letting you know that this is issue is featured on this quarters roadmap. If a PR exists to close the issue a maintainer will review and either make changes directly, or work with the original author to get the contribution merged. If you have written a PR to resolve the issue please ensure the "Allow edits from maintainers" box is checked. Thanks for your patience and we are looking forward to getting this merged soon!

**Comment by Jorge-Rodriguez** [comment]
ping @Avolynsk

**Review by anGie44 — APPROVED** [review]
Thanks so much for this PR @Avolynsk 🚀 Overall, this looked great! I added in some changes to account for the service packages refactor.

Output of acceptance tests:
```
--- PASS: TestAccLambdaLayerVersionPermission_org (24.77s)
--- PASS: TestAccLambdaLayerVersionPermission_basic_byARN (29.05s)
--- PASS: TestAccLambdaLayerVersionPermission_disappears (30.18s)
--- PASS: TestAccLambdaLayerVersionPermission_basic_byName (34.84s)
--- PASS: TestAccLambdaLayerVersionPermission_account (37.62s)
```

**Review by Jorge-Rodriguez — APPROVED** [review]
LGTM

**Pull request merged**
